File: day2/code.ex
defmodule AdventOfCode do
  defp input do
    {:ok, data} = File.read("input.txt")

    data
    |> String.split("\n")
  end

  defp count_depth_and_position(list) when is_list(list) do
    list
    |> Enum.reduce({0, 0}, fn i, acc ->
      case parse_command(i) do
        {:up, value} -> {elem(acc, 0), elem(acc, 1) - value}
        {:down, value} -> {elem(acc, 0), elem(acc, 1) + value}
        {:forward, value} -> {elem(acc, 0) + value, elem(acc, 1)}
        _ -> acc
      end
    end)
    |> multiply()
  end

  defp count_depth_and_position_with_aim(list) when is_list(list) do
    list
    |> Enum.reduce({0, 0, 0}, fn i, acc ->
      case parse_command(i) do
        {:up, value} ->
          {elem(acc, 0), elem(acc, 1), elem(acc, 2) - value}

        {:down, value} ->
          {elem(acc, 0), elem(acc, 1), elem(acc, 2) + value}

        {:forward, value} ->
          {
            elem(acc, 0) + value,
            elem(acc, 1) + elem(acc, 2) * value,
            elem(acc, 2)
          }

        _ ->
          acc
      end
    end)
    |> multiply()
  end

  defp multiply({v1, v2}) when is_integer(v1) and is_integer(v2) do
    v1 * v2
  end

  defp multiply({v1, v2, _}) do
    multiply({v1, v2})
  end

  defp parse_command("forward " <> value) do
    {:forward, String.to_integer(value)}
  end

  defp parse_command("up " <> value) do
    {:up, String.to_integer(value)}
  end

  defp parse_command("down " <> value) do
    {:down, String.to_integer(value)}
  end

  defp parse_command(_) do
    {:unknown}
  end

  def run_part1() do
    input()
    |> count_depth_and_position()
  end

  def run_part2() do
    input()
    |> count_depth_and_position_with_aim()
  end
end

IO.puts("Part1: #{AdventOfCode.run_part1()}")
IO.puts("Part2: #{AdventOfCode.run_part2()}")

File: day2/code.php
<?php

class AdventOfCode
{
    public function part1(): int
    {
        $input = self::getInput();
        $depth = $this->calcDepth($input);
        $position = $this->calcPosition($input);
        return $depth*$position;
    }

    public function part2(): int
    {
        $input = self::getInput();
        $data = $this->calcDepthAndPosition($input);
        return $data[0]*$data[1];
    }

    private function calcDepthAndPosition(array $input): array
    {
        // [position, depth, aim]
        return array_reduce($input, static function ($acc, $item): array {
            if (substr($item, 0, 7) === 'forward') {
                $value = intval(substr($item, 8));
                $acc[0] += $value;
                $acc[1] += $acc[2]*$value;
            } elseif (substr($item, 0, 4) === 'down') {
                $acc[2] += intval(substr($item, 5));
            } elseif (substr($item, 0, 2) === 'up') {
                $acc[2] -= intval(substr($item, 3));
            }
            return $acc;
        }, [0,0,0]);
    }

    private function calcPosition(array $input): int
    {
        return array_reduce($input, static function ($acc, $item): int {
            if (substr($item, 0, 7) === 'forward') {
                $acc += intval(substr($item, 8));
            }
            return $acc;
        }, 0);
    }

    private function calcDepth(array $input): int
    {
        return array_reduce($input, static function ($acc, $item): int {
            if (substr($item, 0, 4) === 'down') {
                $acc += intval(substr($item, 5));
            } elseif (substr($item, 0, 2) === 'up') {
                $acc -= intval(substr($item, 3));
            }
            return $acc;
        }, 0);
    }

    private static function getInput(): array
    {
        return explode("\n", file_get_contents("input.txt"));
    }
}


$aoc = new AdventOfCode();
echo 'Part 1: '.$aoc->part1().PHP_EOL;
echo 'Part 2: '.$aoc->part2().PHP_EOL;
